### PR TITLE
Feature/apl 1259 export data in updater

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
- added export/import 2fa data on startup
- preserve some data by exporting in updater

### CODE REVIEW

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Inject
+    public StartUpdateEventObserver(PropertiesHolder propertiesHolder, OptionDAO optionDAO, ConfigDirProvider configDirProvider) {
+        this.propertiesHolder = Objects.requireNonNull(propertiesHolder, "propertiesHolder is NULL");
+        this.optionDAO = Objects.requireNonNull(optionDAO, "optionDAO is NULL");
+        this.configDirProvider = Objects.requireNonNull(configDirProvider, "configDirProvider is NULL");
+    }
+
+    public void onStartUpdateBefore(
+        @Observes @StartUpdateEvent(StartUpdateEventType.BEFORE_SCRIPT) StartUpdateEventData startUpdateEventData) {
```
Comment (AndrewBoyarsky) [inline apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java]: Maybe rename to @UpdateEvent ?

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        log.debug("onStartUpdateBefore...");
+        // read some properties
+        String nameKey = new String(Base64.getDecoder().decode("YXBsLmFkbWluUGFzc3dvcmQ="));
```
Comment (AndrewBoyarsky) [inline apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java]: YXBsLmFkbWluUGFzc3dvcmQ= looks like better move it to constants and name properly

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        File targetFile = Path.of(configDir, EXPORTED_DATA_FILE_NAME).toFile(); // full target file
+        log.debug("Before saving data to file in config dir = '{}'", targetFile);
+        // read, prepare and write data
+        Properties writeTempProps = new Properties();
+        try (OutputStream output = new FileOutputStream(targetFile, false) ) {
+            // set the properties value
+            log.debug("START write props to file '{}'", targetFile);
+            byte[] nameKeyHash;
+            if (propertyValue != null && !propertyValue.isEmpty()) {
```
Comment (AndrewBoyarsky) [inline apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java]: StringUtil fit ok here

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/service/SecureStorageService.java [hunk 4]
```diff
@@ -7,6 +7,8 @@
 
 public interface SecureStorageService {
 
+    static final String SECURE_STORE_KEY = "secure_store_key";
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/service/SecureStorageService.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthEntity.java [hunk 5]
```diff
@@ -68,4 +68,19 @@ public TwoFactorAuthEntity clone() throws CloneNotSupportedException {
         obj.setSecret(Arrays.copyOf(this.secret, this.secret.length));
         return obj;
     }
+
+    @Override
+    public String toString() {
+        final StringBuffer sb = new StringBuffer("TwoFactorAuthEntity{");
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthEntity.java]: It's a minor issue for our case. StringBuffer is synchronized and does not affect us much.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthRepositoryImpl.java [hunk 6]
```diff
@@ -108,13 +108,35 @@ private boolean executeUpdate(PreparedStatement pstmt) throws SQLException {
...
+    }
+
+    @Override
+    public List<TwoFactorAuthEntity> selectAll() {
+        log.trace("start selectAll 2fa records...");
+        List<TwoFactorAuthEntity> result = new ArrayList<>(10);
+        try (Connection con = db.getConnection();
+            PreparedStatement pstmt = con.prepareStatement(SELECT_QUERY_ALL)) {
+            try (ResultSet rs = pstmt.executeQuery()) {
+                if (rs.next()) {
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthRepositoryImpl.java]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthServiceIntegrationTest.java [hunk 7]
```diff
@@ -152,8 +165,19 @@ public void testConfirmAlreadyConfirmed() throws GeneralSecurityException {
...
     }
+
+    @Test
+    public void testMoveData() throws GeneralSecurityException {
+        TwoFactorAuthTestData td = new TwoFactorAuthTestData();
+        int currentCode = (int) TimeBasedOneTimePasswordUtil.generateCurrentNumber(ACCOUNT3_2FA_SECRET_BASE32);
+        service = new TwoFactorAuthServiceImpl(dbRepository, "test", fileRepository);
+        boolean result = service.attemptMoveDataFromDatabase();
+        assertTrue(result, "Error on moving data from DB into File repo");
+        assertFalse(service.isEnabled(td.ACCOUNT3.getId()));
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthServiceIntegrationTest.java]: done, updated test

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/core/UpdaterCoreImpl.java [hunk 8]
```diff
@@ -96,6 +107,8 @@ public void init(String updateAttachmentFile, boolean debug) {
             if (updateData == null) {
                 LOG.debug("Unable to validate file attachment");
             } else {
+                fireStartUpdateEventMethod(); // fire event for doing data export
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/core/UpdaterCoreImpl.java]: Added one more call

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java]: Renamed

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java]: Done, minor issue

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/StartUpdateEventObserver.java]: No, we don't need, let it be as is.

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/core/UpdaterCoreImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/core/UpdaterCoreImpl.java]: @yuriy-larin-firstbridge I reviewed UpdaterCoreImpl again, no other 'fireStartUpdateEventMethod' calls were added.

[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/core/UpdaterCoreImpl.java]: @AndrewBoyarsky see line 121 on 'files TAB' below

File: apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/UpdateEventObserver.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        } else {
+            log.debug("Exists configDir = '{}'", configDir);
+        }
+        File targetFile = Path.of(configDir, EXPORTED_DATA_FILE_NAME).toFile(); // full target file
+        Properties writeTempProps = new Properties();
+        // check if file has been written previously ans stores something valuable
+        if (targetFile.exists()) {
+            try (InputStream inputStream = new FileInputStream(targetFile)) {
+                writeTempProps.load(inputStream);
+                if (writeTempProps.containsKey(new String(Base64.getDecoder().decode("YXBsLmRwYXM=")))) {
```
Comment (andrii-firstbridge) [inline apl-updater/src/main/java/com/apollocurrency/aplwallet/apl/updater/export/UpdateEventObserver.java]: What does this string mean? "YXBsLmRwYXM="

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TwoFactorAuthServiceImpl.java [hunk 10]
```diff
@@ -150,4 +165,38 @@ private Status2FA process2FAEntity(TwoFactorAuthEntity entity, boolean shouldBeC
...
+        int countProcesses = 0; // store really processed records
+        // loop over all db records if any
+        for (TwoFactorAuthEntity itemRecord: result ) {
+            // check if record was stored as 2fa file before
+            TwoFactorAuthEntity previousValue = targetFileRepository.get(itemRecord.getAccount()); // store into file
+            if (previousValue == null) { // if 2fa file is missing
+                boolean isImported = targetFileRepository.add(itemRecord); // store into file
+                if (!isImported) {
+                    log.error("Error importing 2fa record as File (result={}) = {}", isImported, itemRecord);
+                    isAllImportedIntoFile = isImported; // store false
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TwoFactorAuthServiceImpl.java]: removed

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TwoFactorAuthServiceImpl.java [hunk 11]
```diff
@@ -150,4 +165,38 @@ private Status2FA process2FAEntity(TwoFactorAuthEntity entity, boolean shouldBeC
...
+                    continue; // go to next record, do not delete if record was NOT exported
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TwoFactorAuthServiceImpl.java]: refactored

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthFileSystemRepository.java [hunk 12]
```diff
@@ -108,31 +107,35 @@ public boolean update(TwoFactorAuthEntity entity2FA) {
...
-                LOG.debug("2fa delete json error", e);
+            } catch (IOException e) {
+                log.debug("2fa delete json error", e);
             }
         }
         return false;
     }
+
+    @Override
+    public List<TwoFactorAuthEntity> selectAll() {
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/TwoFactorAuthFileSystemRepository.java]: Done

**Pull request merged**